17016 migrate switch

robertKozik opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
This PR changes the soon-to-be deprecated TouchableOpacity used in Switch component to PressableWithFeedback. In addition this PR fixes two warnings which come from `OpacityView`
### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/17016
PROPOSAL: https://github.com/Expensify/App/issues/17016


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
You can find switch component on the RHN inside preferences tab and on the Timezone initial page

1. Navigate to the page which has switch component
2. Hover and press the switch
3. Verify that switch works fine and is dimming on press


Test changes in PressableWithFeedback:
1.   Add `disabled` prop for `PressableWithFeedback` inside `Switch`
2.  Switch should be disabled and no dimming can be encountered on hover and pressing
3.  No errors on pressing should appear

- [X] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
Same as Tests steps

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
5. Upload an image via copy paste
6. Verify a modal appears displaying a preview of that image
--->
Same as Tests steps

- [X] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/61146594/236581906-bd5334c2-c40a-4bec-8a0a-47a83244ff5b.mov


</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/61146594/236581938-18217e10-ca7d-4eab-89a8-c820401a2e77.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/61146594/236582075-473f4553-71f2-47a1-a8f7-c724bfc670fa.MP4


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/61146594/236581969-fc2448e4-1f2c-4121-821d-688f5f421d85.mov


</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/61146594/236581980-c959f2b0-a49e-48d4-ab79-5681d5c61195.mov


</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/61146594/236581991-96a46906-6b3c-44d9-a814-2bba31faa1cd.mov


</details>

robertKozik committed:
migrate switch to PressableWithFeedback, add accessibility props

robertKozik committed:
fix warnings which come from OpacityView

sobitneupane commented on src/components/Switch.js:
```diff
@@ -1,16 +1,21 @@
     onToggle: PropTypes.func.isRequired,
+
+    /** Accessibility label for the switch */
+    accessibilityLabel: PropTypes.string.isRequired,
```
`accessibilityLabel` is declared as Required but it's value is undefined as it not passed from components using `Switch`.

robertKozik replied:
Filled, good catch
Thanks

sobitneupane replied:
@robertKozik `accessibilityLabel` should be added in `TimezoneInitialPage` as well.

https://github.com/Expensify/App/blob/e52c1e295713a9f67aeb69ba82d011636d8830d6/src/pages/settings/Profile/TimezoneInitialPage.js#L61-L64

sobitneupane reviewed (commented):
When hovered the `Switch` is dimmed which was not the case while using `TochableOpacity`. Is it expected behaviour? Was it discussed before? If we hover the switch when it is toggled off, it is too dark.


https://user-images.githubusercontent.com/25876548/236727571-9af729d4-aad1-432d-9fd9-617984eb2b3d.mov

robertKozik committed:
remove dimming from switch component

robertKozik committed:
add required accessibility labels for all switches

robertKozik commented:
> When hovered the `Switch` is dimmed which was not the case while using `TochableOpacity`. Is it expected behaviour? Was it discussed before? If we hover the switch when it is toggled off, it is too dark.
> 
>  Screen.Recording.2023-05-08.at.09.13.52.mov

PressableWithFeedback has now implemented hover and press dimming out-of-the box as stated inside the [design doc](https://docs.google.com/document/d/1yHHs4IfaMxB2q7eW8Hyt7ew4rNtY6HdduPKdf-Q2ZxE/edit#bookmark=id.z63gelgjqb9o). As there was no dimming for hover before, I've removed it.

robertKozik committed:
Add remaining accessibilityLabel for switch in TimezoneInitialPage.js

robertKozik committed:
fix error when GenericPressable tries to call onPress when disabled

sobitneupane reviewed (commented):
### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/25876548/4a190fd2-06cc-457b-bb7d-5cb466cb871b

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/25876548/91bac7be-69f5-44bf-ba37-505929f2ffff



</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/25876548/e4264362-ecda-4dc6-b9e3-d2b5217b52d8

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/25876548/860db9dd-7dcb-4763-b384-e2fcd90b328d


</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/25876548/195c3b5f-b091-42cf-af37-187d055622f6

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/25876548/644b55bc-e153-41be-afb4-9878eee0d712

</details>

robertKozik commented:
Sorry for the problem, as I pushed additional fix for BaseGenericPressable component – it's not changing behaviour of the Switch, but removes the error when `disabled` state is passed to it (not used in the switch component), so it's worth to include it imo. Sorry for the inconvenience as you already reviewed the changes :/

sobitneupane reviewed (commented):
Differences with staging:

- OnLongPress event toggles the switch on staging which is not the case in this PR.
- On press or On long press, the switch dimmed too much which is not the case in staging. Can we decreasing the dimming for `onpress` and `onlongpress` event as well?

@robertKozik Can you please handle above issues as well?

https://github.com/Expensify/App/assets/25876548/f40e1af1-80af-4227-bc2a-b825a9841362

sobitneupane commented:
No worries @robertKozik. I will re-review it once you fixed [these issues](https://github.com/Expensify/App/pull/18512#pullrequestreview-1418567164).

robertKozik commented on src/components/Pressable/PressableWithFeedback.js:
```diff
@@ -29,7 +29,7 @@ const PressableWithFeedback = forwardRef((props, ref) => {
             {state => (
                 <OpacityView
-                    shouldDim={state.pressed || state.hovered}
+                    shouldDim={state.pressed || state.hovered || false}
```
if both pressed and hovered is undefined it will return undefined and cause warning

deetergp replied:
Ugh. `undefined` Right.

robertKozik committed:
press dim value for switch is set to 0.8

robertKozik committed:
trigger the same callback on press and longPress for switch

robertKozik committed:
Merge remote-tracking branch 'upstream/main' into 17016_migrate_switch

robertKozik committed:
Merge branch 'main' into 17016_migrate_switch

robertKozik committed:
disable dimming of PressableWithFeedback when is disabled

robertKozik commented:
Hello, I changed the dimming value to the one present before this PR – 0.8, hoisted the press callbeck wrapper in order to use it for press and long press, disabled dimming on pressing disabled component, updated the test steps and merged main to this PR

robertKozik committed:
run prettier on changed files

sobitneupane reviewed (dismissed):
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

deetergp reviewed (dismissed):
Looks good, but you've got a merge conflict, unfortunately.

robertKozik committed:
Merge branch 'main' into 17016_migrate_switch

robertKozik commented:
Conflicts has been resolved 👌🏼

deetergp reviewed (approved):
Code looks good and _mostly_ works as advertised. Your testing steps mention that it should dim on over on web, but when I tested it, there was no dimming. I don't think it's a blocker though, since the switches don't currently dim on hover on production. 

All yours, @sobitneupane!

robertKozik commented:
Good catch, I forgot to change the test steps to not involve dimming on hover. I did this exactly to mimic the production environment (https://github.com/Expensify/App/pull/18512#issuecomment-1542814182)

sobitneupane reviewed (approved):
Changes look good and tests well.

cc: @deetergp

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4997424374) to staging by https://github.com/deetergp in version: 1.3.15-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

mvtglobally commented:
@robertKozik @deetergp  We are not seeing Dimming on the switches. Is this something that need to be enabled in the backend? Should it be internal QA? 

https://github.com/Expensify/App/assets/43995119/3c9bbe7f-e213-4b1f-9520-a25005b04abf

robertKozik commented:
@mvtglobally dimming was removed during the PR review process ([comment](https://github.com/Expensify/App/pull/18512#issuecomment-1539146328)). I just saw I forgot to update test steps. Sorry for that

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5011154080) to production by https://github.com/yuwenmemon in version: 1.3.15-12 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
